File: js-src/src/storage/HashTree.ts
import { ChangeType } from "../Change";
import { Store } from "./BaseStore";
import { dumpToDict } from '../libs/Entity';
import { cryptoModule } from "../base-util";
import { getLogger } from "../logging";
import { Delta } from "./Delta";

const log = getLogger('HashTree')

const subtleCrypto = cryptoModule();

function sortObjectProperties(obj: any, depth: number = 1): any {
    if (depth > 3) {
        throw new Error("Depth exceeded: This function supports sorting up to 3 levels deep only.");
    }

    if (Array.isArray(obj)) {
        return obj.map(item =>
            typeof item === 'object' && item !== null ?
                sortObjectProperties(item, depth + 1) :
                item
        );
    }

    if (typeof obj !== 'object' || obj === null) {
        return obj;
    }

    const sorted: { [key: string]: any } = {};
    Object.keys(obj).sort().forEach(key => {
        const value = obj[key];
        sorted[key] = typeof value === 'object' && value !== null ?
            sortObjectProperties(value, depth + 1) :
            value;
    });

    return sorted;
}

function getEntityDataString(entity: any): string {
    const data = dumpToDict(entity);
    const sortedData = sortObjectProperties(data);
    return JSON.stringify(sortedData);
}

enum NodeType {  // The hash struct needs to index multiple trees in the store
    SUPER_ROOT,  // This is the store root
    ROOT,        // Each entity without a parent is a root
    NON_ROOT     // All other entities are non-roots
}

export class HashTreeNode {
    // The store itsel is not an entity, so the root has null id
    tree: HashTree;
    type: NodeType;
    entityId: string; // The super-root has an empty id
    parentId: string = ''; //  root nodes have an empty parent id

    children: { [key: string]: HashTreeNode } = {}; // by entity id
    childrenSorted: HashTreeNode[] = []; // sorted by entity id

    entityDataHash: string = ''; // '' is possible at init or for the super root node (all others are serializable and should be specified on instantiation or at data updates)
    _hash: string = '';

    _hashOutdated: boolean = true;
    _removed: boolean = false;
    _childrenNotSorted: boolean = false;

    constructor(tree: HashTree, entityId: string, parentId: string, entityDataHash: string, type: NodeType) {
        this.tree = tree;
        this.type = type;
        this.entityId = entityId;
        this.parentId = parentId;
        this.entityDataHash = entityDataHash;
    }

    data(): any {
        return {
            entityId: this.entityId,
            parentId: this.parentId,
            entityDataHash: this.entityDataHash,
            hash: this.hash,
            hashOutdated: this.hashOutdated,
            removed: this.removed,
            childrenSortOutdated: this.childrenSortOutdated,
            children: this.childrenSorted.map(child => child.data())
        }
    }

    get hash(): string {
        if (this._hashOutdated) {
            throw Error("Hash requested, but it's outdated.");
        }
        return this._hash;
    }

    get hashOutdated(): boolean {
        return this._hashOutdated;
    }
    setHashOutdated() {
        /**
         * Set hash outdated for this node and all its parents
         */
        this._hashOutdated = true;

        // Set for all parents
        let parent = this.tree.nodes[this.parentId];
        if (parent && !parent.hashOutdated) {
            parent.setHashOutdated();
        }
    }
    get removed(): boolean {
        return this._removed;
    }
    markAsRemoved() {
        this._removed = true;
        this.setHashOutdated();
    }
    get childrenSortOutdated(): boolean {
        return this._childrenNotSorted;
    }
    setChildrenSortOutdated() {
        this._childrenNotSorted = true;
        this.setHashOutdated();
    }
    sortChildren() {
        this.childrenSorted.sort((a, b) => {
            if (a.entityId < b.entityId) {
                return -1;
            }
            if (a.entityId > b.entityId) {
                return 1;
            }
            return 0;
        });
        this._childrenNotSorted = false;
        this.setHashOutdated();
    }

    addChild(child: HashTreeNode) {
        if (child.entityId === ''){
            throw Error("Cannot add a child with empty entity id. This is reserved for the super-root");
        }

        this.children[child.entityId] = child;
        this.childrenSorted.push(child);
        this.setChildrenSortOutdated();
        // this.sortChildren();
    }
    removeChild(child: HashTreeNode) {
        delete this.children[child.entityId];
        let index = this.childrenSorted.indexOf(child);
        if (index === -1) {
            throw Error("Child not found");
        }
        this.childrenSorted.splice(index, 1);
        this.setChildrenSortOutdated(); // not needed if sort is on every insert
    }
    safeForSubtreeRemoval(): boolean {
        if (this.removed && this.childrenSorted.length === 0) {
            return true;
        }
        for (let child of this.childrenSorted) {
            if (!child.removed || !child.safeForSubtreeRemoval()) {
                return false;
            }
        }
        return true;
    }
    async updateHash() {
        if (this.removed) {
            throw Error("Cannot update hash of a removed node");
        }

        if (this.childrenSortOutdated) {
            throw Error("Cannot update hash: children not sorted");
        }

        // Check for any removed children - they should have been cleaned up
        // And update hashes for all children
        for (let child of this.childrenSorted) {
            if (child.removed) {
                throw Error("Cannot update hash: child node marked for removal must be cleaned up first");
            }
            if (child.hashOutdated) {
                await child.updateHash();
            }
        }

        const childHashes = this.childrenSorted.map((child) => child.hash);
        this._hash = await hash(this.entityDataHash, childHashes);
        this._hashOutdated = false;
        this.hash
    }
}

export class HashTree {
    private _tmpSubtrees: { [key: string]: HashTreeNode[] } = {}; // by parent id
    superRoot: HashTreeNode | null = null;
    nodes: { [key: string]: HashTreeNode } = {} // by entity id
    _removedNodeCleanupNeeded: boolean = false; // Handle nodes marked for removal

    constructor() {
        this.createSuperRoot();
    }

    get removedNodeCleanupNeeded(): boolean {
        return this._removedNodeCleanupNeeded;
    }
    cleanUpRemovedNodesLater() {
        this._removedNodeCleanupNeeded = true;
    }

    createSuperRoot() {
        if (!!this.superRoot) {
            throw Error("Cannot have multiple super-root nodes");
        }
        let superRoot = new HashTreeNode(this, '', '', '', NodeType.SUPER_ROOT);
        this.insertNode(superRoot);
    }

    createRoot(entityId: string, entityDataHash: string) {
        if (entityId === '') {
            throw Error("Root nodes must have an entity id");
        }
        let root = new HashTreeNode(this, entityId, '', entityDataHash, NodeType.ROOT);
        this.insertNode(root);
        return root;
    }

    createNonRoot(entityId: string, parentId: string, entityDataHash: string) {
        if (parentId === '') {
            throw Error("Non-root nodes must have a parent id");
        }
        if (entityId === '') {
            throw Error("Non-root nodes must have an entity id");
        }
        let nonRoot = new HashTreeNode(this, entityId, parentId, entityDataHash, NodeType.NON_ROOT);
        this.insertNode(nonRoot);
        return nonRoot;
    }

    insertNode(node: HashTreeNode) {
        if (this.nodes[node.entityId]) {
            throw Error("Node with this entity id already exists");
        }

        let parent = this.nodes[node.parentId];

        if (node.type === NodeType.SUPER_ROOT) { // For adding the super-root
            if (!!this.superRoot){
                throw Error("Cannot have multiple super-root nodes");
            }
            if (parent) {
                throw Error("Wtf. Super-root node cannot have a parent");
            }
            this.superRoot = node;
            this.nodes[node.entityId] = node;

        // Else if the parent is not in the tree, add it to the temporary subtrees
        } else if (!parent) {
            log.info('Parent not found, adding to tmp subtrees', node.parentId)
            if (!this._tmpSubtrees[node.parentId]) {  // Create list if not already present
                this._tmpSubtrees[node.parentId] = [];
            }
            this._tmpSubtrees[node.parentId].push(node);
            return;

        } else {  // If the parent is in the tree, add the node to it
            parent.addChild(node);
            this.nodes[node.entityId] = node;
        }

        // If some of the tmp subtree roots have this node as a parent
        // reattach them properly
        if (this._tmpSubtrees[node.entityId]) {
            log.info('Reattaching tmp subtrees', node.entityId)
            let subtree = this._tmpSubtrees[node.entityId];
            delete this._tmpSubtrees[node.entityId];

            for (let child of subtree) {
                this.insertNode(child); // This will recursively re-check
            }

        }
    }
    removeNode(node: HashTreeNode) {
        log.info('Marking node for removal', node.entityId)
        node.markAsRemoved();
        this.cleanUpRemovedNodesLater();
    }
    parent(node: HashTreeNode): HashTreeNode | null {
        return this.nodes[node.parentId] || null;
    }

    deleteNodesMarkedForRemoval() {
        /** Remove nodes marked for removal */
        log.info('Cleaning up hash tree')
        // Remove nodes marked for removal. First detach from parents, then delete
        let forDeletion: HashTreeNode[] = [];
        for (let node of Object.values(this.nodes)) {
            if (!node.removed){
                continue;
            }

            if(node.safeForSubtreeRemoval()) {
                forDeletion.push(node); // Mark for deletion from the index
                let parent = this.parent(node)
                if (!parent) {
                    throw Error("Unexpected missing parent");
                }
                parent.removeChild(node); // Remove from parent
            } else {
                throw Error("Cannot remove node with children");
            }
        }

        // After all have been removed from the parents - actually delete them
        for (let node of forDeletion) {
            delete this.nodes[node.entityId];
        }

        this._removedNodeCleanupNeeded = false;
    }

    async updateRootHash() {
        if (!this.superRoot) {
            throw Error("Cannot update hash of an empty tree");
        }

        if (this.removedNodeCleanupNeeded) {
            this.deleteNodesMarkedForRemoval();
        }

        // Check for hanging subtrees
        if (Object.keys(this._tmpSubtrees).length > 0) {
            throw Error("Cannot update hash: hanging tmp subtrees");
        }

        // Sort children where needed
        for (let node of Object.values(this.nodes)) {
            if (node.childrenSortOutdated) {
                node.sortChildren();
            }
        }

        await this.superRoot.updateHash();
    }

    rootHash(): string {
        if (!this.superRoot) {
            throw Error("Cannot get hash of an empty tree");
        }
        if (this.removedNodeCleanupNeeded) {
            throw Error("Tree needs cleanup");
        }
        if (this.superRoot.hashOutdated) {
            throw Error("Super-root hash is outdated");
        }
        return this.superRoot.hash;
    }
}


async function hash(data: string, dataForConcat: string[] = []): Promise<string> {
    // Concatenate (child hashes should be sorted by entityId)
    let concatenatedData = data + dataForConcat.join('');

    const encoder = new TextEncoder();
    const dataEncoded = encoder.encode(concatenatedData);
    const hashBuffer: ArrayBuffer = await subtleCrypto.digest("SHA-256", dataEncoded);
    const hashArray = Array.from(new Uint8Array(hashBuffer));
    const hashHex = hashArray.map(byte => byte.toString(16).padStart(2, '0')).join('');

    return hashHex;
}

async function buildSubtree(store: Store, tree: HashTree, startNode: HashTreeNode) {
    const children = Array.from(store.find({ parentId: startNode.entityId }));

    // Add children
    for (let child of children) {
        const data = getEntityDataString(child);
        let childNode = tree.createNonRoot(child.id, child.parentId, await hash(data));
        await buildSubtree(store, tree, childNode)
    }
}

export async function buildHashTree(store: Store): Promise<HashTree> {
    let tree: HashTree = new HashTree();

    // For every root node - add it, and build its subtree
    let rootEntities = Array.from(store.find({ parentId: '' }));
    for (let rootEntity of rootEntities) {
        const data = getEntityDataString(rootEntity);
        let rootNode = tree.createRoot(rootEntity.id, await hash(data));
        await buildSubtree(store, tree, rootNode)
    }
    await tree.updateRootHash();
    return tree
}

export async function updateHashTree(tree: HashTree, store: Store, delta: Delta) {
    // For deletions - remove the node
    for (let change of delta.changes()) {
        let changeType = change.type();
        let entity = store.findOne({ id: change.entityId });

        if (changeType === ChangeType.DELETE) {
            let node = tree.nodes[change.entityId];
            if (!node) {
                throw Error("Cannot delete a node that doesn't exist");
            }
            tree.removeNode(node);

        } else if (changeType === ChangeType.UPDATE) {
            let node = tree.nodes[change.entityId];
            if (!node){
                throw Error("Cannot update a node that doesn't exist");
            }
            if (entity === undefined) {
                throw Error(`Entity not found for the given change ${change.entityId}`);
            }
            const data = getEntityDataString(entity);
            node.entityDataHash = await hash(data);
            node.setHashOutdated();

        } else if (changeType === ChangeType.CREATE) {
            if (entity === undefined) {
                throw Error(`Entity not found for the given change ${change.entityId}`);
            }
            const data = getEntityDataString(entity);
            let hashString = await hash(data);
            let nodeType = entity.parentId === '' ? NodeType.ROOT : NodeType.NON_ROOT;
            let node = new HashTreeNode(tree, change.entityId, entity.parentId, hashString, nodeType);
            tree.insertNode(node);
        } else {
            throw Error("Unexpected change type");
        }
    }
    await tree.updateRootHash();
}

File: py-src/fusion/storage/hash_tree.py
from __future__ import annotations

import hashlib
from typing import Any

from fusion import get_logger
from fusion.libs.canonical_json import canonical_json
from fusion.libs.model import Entity, dump_to_dict
from fusion.storage.base_store import Store
from fusion.storage.change import ChangeTypes
from fusion.storage.delta import Delta

log = get_logger(__name__)


class HangingSubtreesError(Exception):
    pass


def _get_entity_data_string(entity: Entity) -> str:
    data = dump_to_dict(entity)
    return canonical_json(data)


def _hash(data: str, data_for_concat: list[str] | None = None) -> str:
    """SHA-256 hash of data + concatenated child hashes."""
    concatenated = data + "".join(data_for_concat or [])
    return hashlib.sha256(concatenated.encode()).hexdigest()


# Tree structure:
# - Super-root (entity_id='', virtual) anchors all root nodes
# - Root nodes (parent_id='') represent top-level entities (pages)
# - Non-root nodes represent children (notes, arrows)


class HashTreeNode:
    __slots__ = (
        "tree",
        "entity_id",
        "parent_id",
        "children",
        "children_sorted",
        "entity_data_hash",
        "_hash",
        "_hash_outdated",
        "_removed",
        "_children_not_sorted",
    )

    def __init__(
        self,
        tree: HashTree,
        entity_id: str,
        parent_id: str,
        entity_data_hash: str,
    ) -> None:
        self.tree = tree
        self.entity_id = entity_id
        self.parent_id = parent_id
        self.entity_data_hash = entity_data_hash
        self.children: dict[str, HashTreeNode] = {}
        self.children_sorted: list[HashTreeNode] = []
        self._hash: str = ""
        self._hash_outdated: bool = True
        self._removed: bool = False
        self._children_not_sorted: bool = False

    @staticmethod
    def create_super_root(tree: HashTree) -> HashTreeNode:
        return HashTreeNode(tree, "", "", "")

    @staticmethod
    def create_root(
        tree: HashTree, entity_id: str, entity_data_hash: str
    ) -> HashTreeNode:
        if not entity_id:
            raise ValueError("Root nodes must have an entity id")
        return HashTreeNode(tree, entity_id, "", entity_data_hash)

    @staticmethod
    def create(
        tree: HashTree,
        entity_id: str,
        parent_id: str,
        entity_data_hash: str,
    ) -> HashTreeNode:
        if not entity_id:
            raise ValueError("Entity id must not be empty")
        if not parent_id:
            raise ValueError(
                "Non-root nodes must have a parent id (use create_root for roots)"
            )
        return HashTreeNode(tree, entity_id, parent_id, entity_data_hash)

    @property
    def hash(self) -> str:
        if self._hash_outdated:
            raise RuntimeError("Hash requested, but it's outdated.")
        return self._hash

    @property
    def hash_outdated(self) -> bool:
        return self._hash_outdated

    def set_hash_outdated(self) -> None:
        self._hash_outdated = True
        parent = self.tree.nodes.get(self.parent_id)
        if parent and not parent.hash_outdated:
            parent.set_hash_outdated()

    @property
    def removed(self) -> bool:
        return self._removed

    def mark_as_removed(self) -> None:
        self._removed = True
        self.set_hash_outdated()

    @property
    def children_sort_outdated(self) -> bool:
        return self._children_not_sorted

    def set_children_sort_outdated(self) -> None:
        self._children_not_sorted = True
        self.set_hash_outdated()

    def sort_children(self) -> None:
        self.children_sorted.sort(key=lambda n: n.entity_id)
        self._children_not_sorted = False
        self.set_hash_outdated()

    def add_child(self, child: HashTreeNode) -> None:
        if not child.entity_id:
            raise ValueError("Cannot add a child with empty entity id")
        self.children[child.entity_id] = child
        self.children_sorted.append(child)
        self.set_children_sort_outdated()

    def remove_child(self, child: HashTreeNode) -> None:
        del self.children[child.entity_id]
        try:
            self.children_sorted.remove(child)
        except ValueError:
            raise ValueError(
                f"Cannot remove child (entity_id: {child.entity_id}) "
                f"from node (entity_id: {self.entity_id}): child not found"
            )
        self.set_children_sort_outdated()

    def safe_for_subtree_removal(self) -> bool:
        if self.removed and not self.children_sorted:
            return True
        return all(
            child.removed and child.safe_for_subtree_removal()
            for child in self.children_sorted
        )

    def update_hash(self) -> None:
        if self.removed:
            raise RuntimeError("Cannot update hash of a removed node")
        if self.children_sort_outdated:
            raise RuntimeError("Cannot update hash: children not sorted")

        for child in self.children_sorted:
            if child.removed:
                raise RuntimeError(
                    "Cannot update hash: child node marked for removal must be cleaned up first"
                )
            if child.hash_outdated:
                child.update_hash()

        child_hashes = [child.hash for child in self.children_sorted]
        self._hash = _hash(self.entity_data_hash, child_hashes)
        self._hash_outdated = False


class HashTree:
    def __init__(self) -> None:
        self._tmp_subtrees: dict[str, list[HashTreeNode]] = {}
        self.super_root: HashTreeNode | None = None
        self.nodes: dict[str, HashTreeNode] = {}
        self._removed_node_cleanup_needed: bool = False
        self._create_super_root()

    @property
    def removed_node_cleanup_needed(self) -> bool:
        return self._removed_node_cleanup_needed

    def clean_up_removed_nodes_later(self) -> None:
        self._removed_node_cleanup_needed = True

    def _create_super_root(self) -> None:
        if self.super_root is not None:
            raise RuntimeError("Cannot have multiple super-root nodes")
        super_root = HashTreeNode.create_super_root(self)
        self.super_root = super_root
        self.nodes[""] = super_root

    def create_node(
        self, entity_id: str, parent_id: str, entity_data_hash: str
    ) -> HashTreeNode:
        if parent_id == "":
            node = HashTreeNode.create_root(self, entity_id, entity_data_hash)
        else:
            node = HashTreeNode.create(self, entity_id, parent_id, entity_data_hash)
        self.insert_node(node)
        return node

    def insert_node(self, node: HashTreeNode) -> None:
        if node.entity_id in self.nodes:
            raise ValueError(f"Node with entity id {node.entity_id!r} already exists")

        parent = self.nodes.get(node.parent_id)

        if parent is None:
            # Parent not in the tree yet — buffer for later reattachment
            # log.info("Parent not found, adding to tmp subtrees: %s", node.parent_id)
            self._tmp_subtrees.setdefault(node.parent_id, []).append(node)
            return

        parent.add_child(node)
        self.nodes[node.entity_id] = node

        # Reattach any pending children
        if node.entity_id in self._tmp_subtrees:
            log.info("Reattaching tmp subtrees for %s", node.entity_id)
            subtree = self._tmp_subtrees.pop(node.entity_id)
            for child in subtree:
                self.insert_node(child)

    def remove_node(self, node: HashTreeNode) -> None:
        log.info("Marking node for removal: %s", node.entity_id)
        node.mark_as_removed()
        self.clean_up_removed_nodes_later()

    def parent(self, node: HashTreeNode) -> HashTreeNode | None:
        return self.nodes.get(node.parent_id)

    def delete_nodes_marked_for_removal(self) -> None:
        for_deletion: list[HashTreeNode] = []
        for node in self.nodes.values():
            if not node.removed:
                continue
            if node.safe_for_subtree_removal():
                for_deletion.append(node)
                parent = self.parent(node)
                if parent is None:
                    raise RuntimeError("Unexpected missing parent")
                parent.remove_child(node)
            else:
                raise RuntimeError("Cannot remove node with children")

        for node in for_deletion:
            del self.nodes[node.entity_id]

        self._removed_node_cleanup_needed = False

    def update_root_hash(self) -> None:
        if self.super_root is None:
            raise RuntimeError("Cannot update hash of an empty tree")

        if self.removed_node_cleanup_needed:
            self.delete_nodes_marked_for_removal()

        if self._tmp_subtrees:
            raise HangingSubtreesError(
                f"Cannot update hash: hanging tmp subtrees for parent ids: "
                f"{list(self._tmp_subtrees.keys())}"
            )

        # Sort children where needed
        for node in self.nodes.values():
            if node.children_sort_outdated:
                node.sort_children()

        self.super_root.update_hash()

    def root_hash(self) -> str:
        if self.super_root is None:
            raise RuntimeError("Cannot get hash of an empty tree")
        if self.removed_node_cleanup_needed:
            raise RuntimeError("Tree needs cleanup")
        if self.super_root.hash_outdated:
            raise RuntimeError("Super-root hash is outdated")
        return self.super_root.hash


def build_hash_tree(store: Store) -> HashTree:
    """Build a HashTree from the current state of a Store."""
    tree = HashTree()

    # Partition all entities into roots (parent_id="") and children.
    # We iterate all entities because the Python InMemoryStore doesn't index
    # entities with empty parent_id (empty string is falsy).
    all_entities: list[Entity] = list(store.find())
    root_entities = [e for e in all_entities if e.parent_id == ""]
    child_entities = [e for e in all_entities if e.parent_id != ""]

    # Compute all entity data hashes
    hash_by_entity_id: dict[str, str] = {}
    for entity in all_entities:
        data_string = _get_entity_data_string(entity)
        hash_by_entity_id[entity.id] = _hash(data_string)

    # Insert roots first so children find their parents
    for root in root_entities:
        tree.create_node(root.id, "", hash_by_entity_id[root.id])
    for entity in child_entities:
        tree.create_node(entity.id, entity.parent_id, hash_by_entity_id[entity.id])

    tree.update_root_hash()
    return tree


def update_hash_tree(tree: HashTree, store: Store, delta: Delta) -> None:
    """Incrementally update a HashTree after a Delta has been applied to the Store."""
    for change in delta.changes():
        change_type = change.type()
        entity = store.find_one(id=change.entity_id)

        if change_type == ChangeTypes.DELETE:
            node = tree.nodes.get(change.entity_id)
            if node is None:
                raise ValueError(
                    f"Cannot delete a node that doesn't exist: {change.entity_id}"
                )
            tree.remove_node(node)

        elif change_type == ChangeTypes.UPDATE:
            node = tree.nodes.get(change.entity_id)
            if node is None:
                raise ValueError(
                    f"Cannot update a node that doesn't exist: {change.entity_id}"
                )
            if entity is None:
                raise ValueError(f"Entity not found for change {change.entity_id}")
            data_string = _get_entity_data_string(entity)
            node.entity_data_hash = _hash(data_string)
            node.set_hash_outdated()

        elif change_type == ChangeTypes.CREATE:
            if entity is None:
                raise ValueError(f"Entity not found for change {change.entity_id}")
            data_string = _get_entity_data_string(entity)
            hash_string = _hash(data_string)
            tree.create_node(change.entity_id, entity.parent_id, hash_string)

        else:
            raise ValueError(f"Unexpected change type: {change_type}")

    tree.update_root_hash()
